url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/jgraph/example/GraphEdX.java

```diff
@@ -0,0 +1,642 @@
+		GraphConstants.setInset(map, 5);
+		GraphConstants.setGradientColor(map, new Color(200, 200, 255));
+		return map;
+	}
+
+	/**
+	 * Hook from GraphEd to set attributes of a new edge
+	 */
+	public Map createEdgeAttributes() {
+		Map map = super.createEdgeAttributes();
+		// Adds a parallel edge router
+		// GraphConstants.setLineStyle(map, GraphConstants.STYLE_SPLINE);
+		if (GraphConstants.DEFAULTFONT != null) {
+			GraphConstants.setFont(map, GraphConstants.DEFAULTFONT
+					.deriveFont(10f));
+		}
+		return map;
+	}
+
+	/**
+	 * Hook from GraphEd to add action button to the tool bar
+	 */
+	public JToolBar createToolBar() {
+		JToolBar toolbar = super.createToolBar();
+
+		// Collapse
+		collapse = new AbstractAction() {
+			public void actionPerformed(ActionEvent e) {
+				graph.getGraphLayoutCache().setVisible(graph.getSelectionCells(), false);
+			}
+		};
+		URL url = getClass().getClassLoader().getResource(
+				"org/jgraph/example/resources/collapse.gif");
+		collapse.putValue(Action.SMALL_ICON, new ImageIcon(url));
+		collapse.setEnabled(false);
+		toolbar.add(collapse);
+
+		// Expand
+		expand = new AbstractAction() {
+			public void actionPerformed(ActionEvent e) {
+				graph.getGraphLayoutCache().expand(graph.getSelectionCells());
+			}
+		};
+		url = getClass().getClassLoader().getResource(
+				"org/jgraph/example/resources/expand.gif");
+		expand.putValue(Action.SMALL_ICON, new ImageIcon(url));
+		expand.setEnabled(false);
+		toolbar.add(expand);
+
+		// ExpandAll
+		expandAll = new AbstractAction() {
+			public void actionPerformed(ActionEvent e) {
+				Object[] allCells = graph.getDescendants(graph.getRoots());
+				Map nested = new Hashtable();
+				for (int i=0; i < allCells.length; i++) {
+					Map attributeMap = new Hashtable();
+					GraphConstants.setForeground(attributeMap, Color.BLACK);
+					nested.put(allCells[i], attributeMap);
+				}
+				graph.getModel().edit(nested, null, null, null);
+			}
+		};
+		url = getClass().getClassLoader().getResource(
+				"org/jgraph/example/resources/expandAll.gif");
+		expandAll.putValue(Action.SMALL_ICON, new ImageIcon(url));
+		expandAll.setEnabled(false);
+		toolbar.add(expandAll);
+		return toolbar;
+	}
+
+	public void serializeGraph() {
+		int returnValue = JFileChooser.CANCEL_OPTION;
+		initFileChooser();
+		returnValue = fileChooser.showSaveDialog(graph);
+		if (returnValue == JFileChooser.APPROVE_OPTION) {
+			Container parent = graph.getParent();
+			BasicMarqueeHandler marquee = graph.getMarqueeHandler();
+			graph.setMarqueeHandler(null);
+			try {
+				// Serializes the graph by removing it from the component
+				// hierarchy and removing all listeners from it. The marquee
+				// handler, begin an inner class of GraphEd, is not marked
+				// serializable and will therefore not be stored. This must
+				// be taken into account when deserializing a graph.
+				uninstallListeners(graph);
+				parent.remove(graph);
+				ObjectOutputStream out = new ObjectOutputStream(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at lines 315, 316, 318, 319. The resource is closed at line 320. There are other execution paths that do not contain closure statements, for example, when ObjectOutputStream.writeObject throws an exception. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).